### ISSUE TITLE
Allowing ABCI Queries while computing blocks

### ISSUE THREAD

**Issue by michaelfig** [tendermint/tendermint#6899]
## Summary

Allow (at the very least) ABCI QuerySync to run in parallel with AppHash computation.

## Problem Definition

The mutex in `local_client.go` is preventing Cosmos SDK from replying to ABCI QuerySync requests while the chain's state machine is computing the AppHash.  This is bad for scalability for chains with high block compute utilisation whose RPC fullnodes also need to provide low-latency RPC queries.

### Agoric's conundrum

The [Agoric](https://agoric.com) Cosmos-SDK-based chain is intended to provide hardened JS smart contract services that expects to have high utilisation.  We have mechanisms for postponing and scheduling work for future blocks, but we do much more computation per block (say 8-10 seconds to compute the AppHash for the next block) than seems typical of other Cosmos chains.  Additionally, our chain is part of a distributed object system that will have many automated clients making RPC requests (whether submitting transactions or querying state to catch up more quickly than relying exclusively on events).

The aggressive serialisation of ABCI causes our RPC nodes to be unresponsive while computing their AppHashes. This leads to a "thundering herd" problem where clients making queries are blocked while the AppHash is being computed (for 8 seconds). There's a brief moment where voting happens and several (I count typically 25) RPC queries are served, until the cycle begins again and more queries pile up.  This results in a block or two that has high utilisation, followed by several empty blocks while the clients compete to read their state and send new transactions. 

### Cosmos evolution

In #6048 there was some discussion around lock contention and query performance.  There was even a question by @alexanderbez in https://github.com/tendermint/tendermint/issues/6048#issuecomment-784251474 why info and query need to be serialised at all.

Recently, IAVL has been improved and Cosmos SDK is removing its own global ABCI lock around its GRPC Query handlers cosmos/cosmos-sdk#10045.

I believe the rationale for having a single serialised lock (which was allegedly to aid the Cosmos SDK) is obsolete.

With some careful implementation and testing, much more fine-grained locking (if any) would make queries responsive and serviceable by some threads, even while other threads are fully loaded computing the AppHash.

## Proposal

We need a careful analysis of what the actual locking requirements are for the `local_node.go`.

At the very least, I've tested a naive proof-of-concept change to add a separate `queryMtx` that is independent from the existing `mtx` (**not** a RWLock, since that would still block during our AppHash calculation in EndBlock), and to use only the `queryMtx` for `QuerySync()` method.  That's enough to unblock the RPC nodes and free them up to do their jobs at every phase of the ABCI cycle.    I can achieve a steady 100 qps between `curl` and a single Agoric RPC node running on my laptop, even while EndBlock is busy for 8 seconds at a time.

Some consequences of that PoC are separate queries are still serialized.  Folks with deeper Tendermint and Cosmos knowledge need to help decide the best path forward.  I'm reluctant to propose anything except for this strawman because I lack that knowledge.

Thanks for Tendermint!

Attn: @dtribble, @JimLarson, @warner, @zmanian

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by michaelfig** [comment]
I tried removing all locking for `CheckTxSync`, `InfoSync`, and `QuerySync`, and got the following results:

```
ab -n 16000 -c 5 http://localhost:1317/bank/balances/agoric13eh5hru4rh4jkw4khqzrdwgznsj2lf06jqpa4s
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1600 requests
Completed 3200 requests
Completed 4800 requests
Completed 6400 requests
Completed 8000 requests
Completed 9600 requests
Completed 11200 requests
Completed 12800 requests
Completed 14400 requests
Completed 16000 requests
Finished 16000 requests


Server Software:        
Server Hostname:        localhost
Server Port:            1317

Document Path:          /bank/balances/agoric13eh5hru4rh4jkw4khqzrdwgznsj2lf06jqpa4s
Document Length:        139 bytes

Concurrency Level:      5
Time taken for tests:   7.945 seconds
Complete requests:      16000
Failed requests:        0
Total transferred:      7792000 bytes
HTML transferred:       2224000 bytes
Requests per second:    2013.82 [#/sec] (mean)
Time per request:       2.483 [ms] (mean)
Time per request:       0.497 [ms] (mean, across all concurrent requests)
Transfer rate:          957.75 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    1  16.7      0    1221
Processing:     1    2  29.2      1    1464
Waiting:        0    2  26.8      1    1464
Total:          1    2  33.7      2    1465

Percentage of the requests served within a certain time (ms)
  50%      2
  66%      2
  75%      2
  80%      2
  90%      2
  95%      2
  98%      3
  99%      3
 100%   1465 (longest request)
```

While this HTTP API query was executing:
1. The chain was running an additional 6s of busy-looping during EndBlock
2. with 5s of commit_timeout
3. an `ag-solo` client was simultaneously interacting with the VM (using an unoptimised communication path consisting of at least one query and transaction on every block)

Unfortunately, (and reproducibly), if I immediately start another benchmark after the above 16000 queries, then I get a hang after ~300 more queries:

```
ab -n 3000 -c 5 http://localhost:1317/bank/balances/agoric13eh5hru4rh4jkw4khqzrdwgznsj2lf06jqpa4s
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 300 requests
apr_socket_recv: Operation timed out (60)
Total of 343 requests completed
```

I suspect there is some rate-limiting logic in place, but I didn't find a way to tune it in the `config.toml` or `app.toml`.  If I rerun the test a few seconds later, it succeeds with the same pattern again (a maximum of ~16300 quick requests before a hang and timeout).

**Comment by michaelfig** [comment]
For comparison, here is the same benchmark as above, but with stock Tendermint v0.34.12:

TL;DR: without patch = 84 qps, with patch = 2000 qps

```
ab -n 16000 -c 5 http://localhost:1317/bank/balances/agoric13txu6sva3uhnnlhzgxatpqt6h0htwaqz4qyj6k
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1600 requests
Completed 3200 requests
Completed 4800 requests
Completed 6400 requests
Completed 8000 requests
Completed 9600 requests
Completed 11200 requests
Completed 12800 requests
Completed 14400 requests
Completed 16000 requests
Finished 16000 requests


Server Software:        
Server Hostname:        localhost
Server Port:            1317

Document Path:          /bank/balances/agoric13txu6sva3uhnnlhzgxatpqt6h0htwaqz4qyj6k
Document Length:        27 bytes

Concurrency Level:      5
Time taken for tests:   191.856 seconds
Complete requests:      16000
Failed requests:        0
Total transferred:      5984000 bytes
HTML transferred:       432000 bytes
Requests per second:    83.40 [#/sec] (mean)
Time per request:       59.955 [ms] (mean)
Time per request:       11.991 [ms] (mean, across all concurrent requests)
Transfer rate:          30.46 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.2      0       5
Processing:     1   60 585.0      2    6424
Waiting:        0   59 585.0      2    6424
Total:          1   60 585.0      3    6424

Percentage of the requests served within a certain time (ms)
  50%      3
  66%      3
  75%      3
  80%      3
  90%      4
  95%      5
  98%      8
  99%     31
 100%   6424 (longest request)
+ ab -n 3000 -c 5 http://localhost:1317/bank/balances/agoric13txu6sva3uhnnlhzgxatpqt6h0htwaqz4qyj6k
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 300 requests
Completed 600 requests
Completed 900 requests
Completed 1200 requests
Completed 1500 requests
Completed 1800 requests
Completed 2100 requests
Completed 2400 requests
Completed 2700 requests
Completed 3000 requests
Finished 3000 requests


Server Software:        
Server Hostname:        localhost
Server Port:            1317

Document Path:          /bank/balances/agoric13txu6sva3uhnnlhzgxatpqt6h0htwaqz4qyj6k
Document Length:        27 bytes

Concurrency Level:      5
Time taken for tests:   40.750 seconds
Complete requests:      3000
Failed requests:        0
Total transferred:      1122000 bytes
HTML transferred:       81000 bytes
Requests per second:    73.62 [#/sec] (mean)
Time per request:       67.917 [ms] (mean)
Time per request:       13.583 [ms] (mean, across all concurrent requests)
Transfer rate:          26.89 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.4      0       8
Processing:     1   67 649.9      2    8422
Waiting:        0   67 649.9      2    8422
Total:          1   68 649.9      3    8423

Percentage of the requests served within a certain time (ms)
  50%      3
  66%      3
  75%      3
  80%      4
  90%      4
  95%      6
  98%      9
  99%   6006
 100%   8423 (longest request)
```

**Comment by zmanian** [comment]
Curious what happens if you run the Tendermint and SDk tests against the fine grained lock branch

**Comment by michaelfig** [comment]
> Curious what happens if you run the Tendermint and SDk tests against the fine grained lock branch

Results look like:
- tendermint/tendermint#6900 - one "data race" detection failure against tendermint:master
- cosmos/cosmos-sdk#10089 - failing horribly against cosmos-sdk:master

**Comment by michaelfig** [comment]
> Curious what happens if you run the Tendermint and SDk tests against the fine grained lock branch

Oh, I forgot to update `go.mod` in cosmos/cosmos-sdk#10089 !  Now it's much better:

Results look like:
- tendermint/tendermint#6900, against `tendermint:master` as of 1f8bb74bba0523aacdb76ed140339c48078c669b
  - one "data race" failure, which is probably explicitly testing for the old behaviour
  - everything else is passing!
- cosmos/cosmos-sdk#10089, against `cosmos:master` as of 78b151dd972f0d0041df5e7b1a0bb0b153f51a7e
  - one `cosmovisor` test failure, apparently unrelated (I reported it in cosmos/cosmos-sdk#10090)
  - test-sim-after-import is taking a long time and I got impatient
  - everything else is passing!

**Comment by marbar3778** [comment]
Have you tried testing against https://github.com/cosmos/cosmos-sdk/pull/10045. This way we get rid of the need to go through tendermint. This pr would be less breaking than this one, but at the same time, i think only the sdk uses the abci query and not other frameworks

It makes sense that we remove the mutex from the abci client in general all at once and pass the handling of it to the app.

**Comment by zmanian** [comment]
@marbar3778 Afaik the agoric app is still stalling with #10045 merged in.

One concern I have with this line of work is that we don't want a query to ever hit a partially processed block. 

Agoric is doing something at the app level I think to make sure that queries are directed against the last fully processed block but there might need to be something at the ABCI layer ensure this.

**Comment by marbar3778** [comment]
is agoric querying tendermint blocks or sdk related data? maybe there is something i can read to catch myself up on this?

**Comment by michaelfig** [comment]
> is agoric querying tendermint blocks or sdk related data?

Querying SDK related data, and also submitting transactions to the mempool.  We can migrate most of the queries into TM events, but any RPC interactions pile up during the BeginBlock->DeliverTx*->EndBlock cycle when the VM is busy (because of the interaction of Query, CheckTx with the mtx that is held) but there is no voting.  Then all the RPCs are suddenly served in the ~1s of voting time (thundering herd), until the next cycle consumes the VM again.

> maybe there is something i can read to catch myself up on this?

I wish there was better documentation, but it's currently in flux.  Our latest plan is Agoric/agoric-sdk#3763.  I've put some of the relevant background in that issue, feel free to ask for more context if you need.

**Comment by michaelfig** [comment]
> One concern I have with this line of work is that we don't want a query to ever hit a partially processed block.

Completely agreed!

> Agoric is doing something at the app level I think to make sure that queries are directed against the last fully processed block
Actually, we aren't doing anything yet.  Still in the experimental phase, and that's one of the concerns that @warner raised.

> but there might need to be something at the ABCI layer ensure this.

That would be great!

**Comment by michaelfig** [comment]
> This pr would be less breaking than this one, but at the same time, i think only the sdk uses the abci query and not other frameworks

> It makes sense that we remove the mutex from the abci client in general all at once and pass the handling of it to the app.

I appreciate that the default locking behaviour of `localClient` can be of benefit to people who are not trying to squeeze every ounce of performance out of their chain.

I've come to understand that there's a simple compatible way forward without breaking any other clients that might rely on such locking.  Cosmos-SDK currently use `proxy.NewLocalClientConfigCreator`.  If we made a new kind of `localClient`, say `unlockedLocalClient`, and had `proxy.NewUnlockedLocalClientCreator`, then Cosmos-SDK could opt into the new behaviour without breaking existing clients.  For testing, the Agoric fork of Cosmos-SDK could do that while we work on the feature and before it's used by other chains.

**Comment by marbar3778** [comment]
> Querying SDK related data, and also submitting transactions to the mempool. We can migrate most of the queries into TM events, but any RPC interactions pile up during the BeginBlock->DeliverTx*->EndBlock cycle when the VM is busy (because of the interaction of Query, CheckTx with the mtx that is held) but there is no voting. Then all the RPCs are suddenly served in the ~1s of voting time (thundering herd), until the next cycle consumes the VM again.

I feel like I missing something in the process of the application. If we remove query routing through tendermint this shouldn't be an issue, if the app is querying tendermint related data (block, header, etc...) then this could make sense, but this wouldn't be routed through the query endpoint.

**Comment by michaelfig** [comment]
> > Querying SDK related data, and also submitting transactions to the mempool. We can migrate most of the queries into TM events, but any RPC interactions pile up during the BeginBlock->DeliverTx*->EndBlock cycle when the VM is busy (because of the interaction of Query, CheckTx with the mtx that is held) but there is no voting. Then all the RPCs are suddenly served in the ~1s of voting time (thundering herd), until the next cycle consumes the VM again.

> I feel like I missing something in the process of the application. If we remove query routing through tendermint this shouldn't be an issue,

I found for both `appd query --node=xxx:26657 ...` and `curl http://xxx:1317/...` that the queries were blocked during computation (EndBlock), even after the removal of the GRPC locking.  I may have also seen interference with the submission of new transactions to the mempool, but I'd have to confirm.

> if the app is querying tendermint related data (block, header, etc...) then this could make sense, but this wouldn't be routed through the query endpoint.

Does the above help clarify?

**Comment by zmanian** [comment]
Yeah both 26657 and 1317 queries those queries are expected to be locked still.

Queries on 9090 should not be blocked anymore

**Comment by marbar3778** [comment]
ohhhh, the app still queries through tendermint for something's. I forgot about this part of the cli. hrmmm, does the sdk need to expose this functionality tho? can it just get rid of querying application data through tendermint?

Im def in favor of the change you are proposing, just trying to better understand the issues in the sdk.

**Comment by ebuchman** [comment]
iirc the local client is using the mutex on everything for simplicity, though it's clearly a bottleneck and should be addressed.

One question though, I believe the non-local abci client does support some concurrency here. Have you tried running the Agoric app out-of-process with Tendermint? That may not be a realistic solution for operational reasons but it could be illuminating, ie. the local client could be adapted to emulate the remote one and not use a global mutex on all abci connections

**Comment by cmwaters** [comment]
For me, this issue seems to circle two different aspects which we may want to consider as potential candidates alongside the ABCI++ work for the 0.36 cycle:
- The `Query` method within ABCI was perhaps added to help applications leverage Tendermint's existing RPC server as a means of querying the state machine. At the time, it might have been thought that the application would be something minimal and wrapped around by Tendermint. However, nowadays the SDK provides a lot of this tooling. By having both gRPC and REST servers that handle application querying, this no longer needs to be done via the Tendermint RPC. Thus, we should remove `Query` and the respective RPC endpoint. 
- For both safety and simplicity, the local client has a mutex on all methods. For certain applications, this locking may not be necessary and cause things to slow down. It should really be the application that has control over how they want to approach concurrency/serialization. The shifting of responsibility to the application of course needs to be clearly communicated hence it's probably best to do that alongside an already breaking change like ABCI++.

**Comment by cmwaters** [comment]
I'm going to close this issue in favor of #7073
